common/src/address.rs

```diff
@@ -0,0 +1,163 @@
+use serde::{Deserialize, Serialize};
+use std::net::{Ipv6Addr, SocketAddrV6};
+
+pub const AZ_PREFIX: u8 = 48;
+pub const RACK_PREFIX: u8 = 56;
+pub const SLED_PREFIX: u8 = 64;
+
+/// The amount of redundancy for DNS servers.
+///
+/// Must be less than MAX_DNS_REDUNDANCY.
+pub const DNS_REDUNDANCY: usize = 1;
+/// The maximum amount of redundancy for DNS servers.
+///
+/// This determines the number of addresses which are
+/// reserved for DNS servers.
+pub const MAX_DNS_REDUNDANCY: usize = 5;
+
+pub const DNS_PORT: u16 = 53;
+pub const DNS_SERVER_PORT: u16 = 5353;
+pub const SLED_AGENT_PORT: u16 = 12345;
+
+/// Represents a subnet which may be used for contacting DNS services.
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct DnsSubnet {
+    network: Ipv6Network,
+}
+
+impl DnsSubnet {
+    /// Returns the DNS server address within the subnet.
+    ///
+    /// This is the first address within the subnet.
+    pub fn dns_address(&self) -> Ipv6Network {
+        let mut iter = self.network.iter();
+        let _anycast_ip = iter.next().unwrap();
+        let dns_ip = iter.next().unwrap();
+        Ipv6Network::new(dns_ip, SLED_PREFIX).unwrap()
+    }
+
+    /// Returns the address which the Global Zone should create
+    /// to be able to contact the DNS server.
+    ///
+    /// This is the second address within the subnet.
+    pub fn gz_address(&self) -> Ipv6Network {
+        let mut iter = self.network.iter();
+        let _anycast_ip = iter.next().unwrap();
+        let _dns_ip = iter.next().unwrap();
+        Ipv6Network::new(iter.next().unwrap(), SLED_PREFIX).unwrap()
+    }
+}
+
+/// A wrapper around an IPv6 network, indicating it is a "reserved" rack
+/// subnet which can be used for AZ-wide services.
+#[derive(Debug, Clone)]
+pub struct ReservedRackSubnet(pub Ipv6Network);
+
+impl ReservedRackSubnet {
+    /// Returns the subnet for the reserved rack subnet.
+    pub fn new(subnet: Ipv6Network) -> Self {
+        let net = Ipv6Network::new(subnet.network(), AZ_PREFIX).unwrap();
+        ReservedRackSubnet(
+            Ipv6Network::new(net.network(), RACK_PREFIX).unwrap(),
+        )
+    }
+
+    /// Given a particular rack subnet, return the DNS addresses.
+    ///
+    /// These addresses will come from the first [`DNS_REDUNDANCY`] `/64s` of the
+    /// [`RACK_PREFIX`] subnet.
+    pub fn get_dns_subnets(&self) -> Vec<DnsSubnet> {
+        assert_eq!(self.0.prefix(), RACK_PREFIX);
+
+        (0..DNS_REDUNDANCY)
+            .map(|idx| {
+                let network =
+                    get_64_subnet(self.0, u8::try_from(idx + 1).unwrap());
+
+                DnsSubnet { network }
+            })
+            .collect()
+    }
+}
+
+/// Return the sled agent address for a subnet.
+///
+/// This address will come from the first address of the [`SLED_PREFIX`] subnet.
+pub fn get_sled_address(sled_subnet: Ipv6Network) -> SocketAddrV6 {
+    assert_eq!(sled_subnet.prefix(), SLED_PREFIX);
```

Comment:
[Const generics](https://blog.rust-lang.org/2021/02/26/const-generics-mvp-beta.html) could give us a nice compile-time safety win here, turning these asserts into compile-time checks. See this [playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e618e0849ef73f683259aee6e7db4b48) for creating a `Ipv6Prefix::<N>` type.